app/controllers/test/oidc_test_controller.rb

```diff
@@ -0,0 +1,154 @@
+require './spec/support/oidc_auth_helper'
+module Test
+  class OidcTestController < ApplicationController
+    include OidcAuthHelper
+
+    BIOMETRIC_REQUIRED = 'biometric-comparison-required'
+
+    def initialize
+      @client_id = 'urn:gov:gsa:openidconnect:sp:test'
+      @update_redirect_uri = true
+      super
+    end
+
+    def index
+      # default to require
+      @start_url_selfie = "#{test_oidc_auth_request_url}?ial=biometric-comparison-required"
+      @start_url_ial2 = "#{test_oidc_auth_request_url}?ial=2"
+      @start_url_ial1 = "#{test_oidc_auth_request_url}?ial=1"
+      update_service_provider if @update_redirect_uri
+    end
+
+    def auth_request
+      ial = prepare_step_up_flow(ial: params[:ial])
+
+      idp_url = authorization_url(
+        ial: ial,
+        aal: params[:aal],
+      )
+
+      Rails.logger.info("Redirecting to #{idp_url}")
+
+      redirect_to(idp_url)
+    end
+
+    def auth_result
+      redirect_to('/')
+    end
+
+    def logout
+      redirect_to(logout_uri)
+    end
+
+    def authorization_url(ial:, aal: nil)
+      authorization_endpoint = openid_configuration[:authorization_endpoint]
+      params = ial2_params(
+        client_id: client_id,
+        acr_values: acr_values(ial: ial, aal: aal),
+        biometric_comparison_required: ial == BIOMETRIC_REQUIRED,
+        state: random_value,
+        nonce: random_value,
+      )
+      request_params = params.merge(
+        {
+          scope: scopes_for(ial),
+          redirect_uri: test_oidc_auth_result_url,
+        },
+      ).compact.to_query
+      "#{authorization_endpoint}?#{request_params}"
+    end
+
+    def prepare_step_up_flow(ial:)
+      if ial == 'step-up'
+        ial = '1'
+      end
+      ial
+    end
+
+    def scopes_for(ial)
+      case ial
+      when '0'
+        'openid email social_security_number'
+      when '1', nil
+        'openid email'
+      when '2', BIOMETRIC_REQUIRED
+        'openid email profile social_security_number phone address'
+      else
+        raise ArgumentError.new("Unexpected IAL: #{ial.inspect}")
+      end
+    end
+
+    def acr_values(ial:, aal:)
+      values = []
+
+      values << {
+        '0' => 'http://idmanagement.gov/ns/assurance/ial/0',
+        nil => 'http://idmanagement.gov/ns/assurance/ial/1',
+        '' => 'http://idmanagement.gov/ns/assurance/ial/1',
+        '1' => 'http://idmanagement.gov/ns/assurance/ial/1',
+        '2' => 'http://idmanagement.gov/ns/assurance/ial/2',
+        'biometric-comparison-required' => 'http://idmanagement.gov/ns/assurance/ial/2',
+      }[ial]
+
+      values << {
+        '2' => 'http://idmanagement.gov/ns/assurance/aal/2',
+        '2-phishing_resistant' => 'http://idmanagement.gov/ns/assurance/aal/2?phishing_resistant=true',
+        '2-hspd12' => 'http://idmanagement.gov/ns/assurance/aal/2?hspd12=true',
+      }[aal]
+
+      values.compact.join(' ')
```

Comment:
I think the `values << {}[]` syntax is a little sneaky, at first skim it looks like we're adding an entire hash to the array.

How about:

```suggestion
      ial_value = {
        '0' => 'http://idmanagement.gov/ns/assurance/ial/0',
        nil => 'http://idmanagement.gov/ns/assurance/ial/1',
        '' => 'http://idmanagement.gov/ns/assurance/ial/1',
        '1' => 'http://idmanagement.gov/ns/assurance/ial/1',
        '2' => 'http://idmanagement.gov/ns/assurance/ial/2',
        'biometric-comparison-required' => 'http://idmanagement.gov/ns/assurance/ial/2',
      }[ial]

      aal_value = {
        '2' => 'http://idmanagement.gov/ns/assurance/aal/2',
        '2-phishing_resistant' => 'http://idmanagement.gov/ns/assurance/aal/2?phishing_resistant=true',
        '2-hspd12' => 'http://idmanagement.gov/ns/assurance/aal/2?hspd12=true',
      }[aal]

      [ial_value, aal_value].compact.join(' ')
```

Reply:
changed.